File: src/ssvep/fbcca-js/test_fbcca.js
// ------------------------------------------------ !NOT! CATERING FOR RESAMPLING/DOWN SAMPLING ------------------------------------

// const math = require('mathjs');
// const { Configuration } = require('./config.js');
// const { filterbank } = require('./filterbank.js');
// const { canoncorr } = require('./cca.js');

// function test_fbcca(eeg, listFreqs) {
//     if (!eeg || !listFreqs) {
//         throw new Error('Not enough input arguments.');
//     }

//     // Filter bank coefficients
//     const fbCoefs = Array.from({ length: Configuration.subBands }, (_, i) => Math.pow(i + 1, -1.25) + 0.25);

//     const numSmpls = eeg[0].length;
//     const numTargs = listFreqs.length;
//     const yRef = ccaReference(listFreqs, numSmpls);

//     const r = Array(Configuration.subBands).fill(null).map(() => Array(numTargs).fill(0));

//     for (let fbI = 0; fbI < Configuration.subBands; fbI++) {
//         const testData = filterbank(eeg, Configuration.samplingRate, fbI + 1);
//         for (let classI = 0; classI < numTargs; classI++) {
//             const refData = yRef[classI];
//             // Perform canonical correlation analysis
//             const r_tmp = canoncorr(testData, refData);

//             r[fbI][classI] = r_tmp[0];
//         }
//     }

//     // Compute rho and find the max correlation
//     const rho = math.multiply(fbCoefs, r);
//     const correlation = Math.max(...rho);
//     const tau = rho.indexOf(correlation);

//     // Check correlation threshold
//     const estimatedLabel = correlation < Configuration.correlationThreshold ? Configuration.idleStateLabel : tau;

//     return estimatedLabel;
// }

// function ccaReference(listFreqs, numSmpls) {
//     if (!listFreqs || !numSmpls) {
//         throw new Error('Not enough input arguments.');
//     }

//     const numFreqs = listFreqs.length;
//     const tidx = Array.from({ length: numSmpls }, (_, i) => (i + 1) / Configuration.samplingRate);
//     const yRef = Array(numFreqs).fill(null).map(() => Array(2 * Configuration.harmonics).fill(null).map(() => Array(numSmpls).fill(0)));

//     for (let freqI = 0; freqI < numFreqs; freqI++) {
//         const tmp = [];
//         for (let harmI = 1; harmI <= Configuration.harmonics; harmI++) {
//             const stimFreq = listFreqs[freqI];
//             tmp.push(tidx.map(t => Math.sin(2 * Math.PI * t * harmI * stimFreq)));
//             tmp.push(tidx.map(t => Math.cos(2 * Math.PI * t * harmI * stimFreq)));
//         }
//         yRef[freqI] = tmp;
//     }

//     return yRef;
// }

// module.exports = { test_fbcca };





// ------------------------------------------------ CATERING FOR RESAMPLING/DOWN SAMPLING ------------------------------------
const math = require('mathjs');
const { fbccaConfiguration } = require('./fbcca_config.js');
const { filterbank } = require('./filterbank.js');
const { canoncorr } = require('./cca.js');

function test_fbcca(eeg, listFreqs) {
    if (!eeg || !listFreqs) {
        throw new Error('Not enough input arguments.');
    }

    // Filter bank coefficients
    const fbCoefs = Array.from({ length: fbccaConfiguration.subBands }, (_, i) => Math.pow(i + 1, -1.25) + 0.25);

    // Get number of samples after downsampling (assuming filterbank handles this)
    const testDataExample = filterbank(eeg, fbccaConfiguration.samplingRate, 1);  // Sub-band 1
    const numSmplsResampled = testDataExample[0].length;

    // Generate reference signals with matching length and 256 Hz sampling rate
    const yRef = ccaReference(listFreqs, numSmplsResampled, 256);  // <-- fixed sampling rate

    const r = Array(fbccaConfiguration.subBands).fill(null).map(() => Array(listFreqs.length).fill(0));

    for (let fbI = 0; fbI < fbccaConfiguration.subBands; fbI++) {
        const testData = filterbank(eeg, fbccaConfiguration.samplingRate, fbI + 1);  // Includes downsampling to 256 Hz
        for (let classI = 0; classI < listFreqs.length; classI++) {
            const refData = yRef[classI];

            const r_tmp = canoncorr(testData, refData);  // Make sure this returns an array of canonical correlations

            r[fbI][classI] = r_tmp[0];  // Only use the first canonical correlation
        }
    }

    // Weighted sum of correlations
    const rho = math.multiply(fbCoefs, r);
    const correlation = Math.max(...rho);
    const tau = rho.indexOf(correlation);

    const estimatedLabel = correlation < fbccaConfiguration.correlationThreshold
        ? fbccaConfiguration.idleStateLabel
        : tau;

    return estimatedLabel;
}

function ccaReference(listFreqs, numSmpls, fs = 256) {
    if (!listFreqs || !numSmpls) {
        throw new Error('Not enough input arguments.');
    }

    const numFreqs = listFreqs.length;
    const tidx = Array.from({ length: numSmpls }, (_, i) => i / fs);  // 0-based for JS
    const yRef = Array(numFreqs).fill(null).map(() =>
        Array(2 * fbccaConfiguration.harmonics).fill(null).map(() => Array(numSmpls).fill(0))
    );

    for (let freqI = 0; freqI < numFreqs; freqI++) {
        const tmp = [];
        const stimFreq = listFreqs[freqI];
        for (let harmI = 1; harmI <= fbccaConfiguration.harmonics; harmI++) {
            tmp.push(tidx.map(t => Math.sin(2 * Math.PI * t * harmI * stimFreq)));
            tmp.push(tidx.map(t => Math.cos(2 * Math.PI * t * harmI * stimFreq)));
        }
        yRef[freqI] = tmp;
    }

    return yRef;
}

module.exports = { test_fbcca };

File: src/ssvep/fbcca-py/test_fbcca.py
import numpy as np
from fbcca_config_service import fbcca_config
from sklearn.cross_decomposition import CCA
from filterbank import filterbank  # Make sure it downsample to 256 Hz internally

def test_fbcca(eeg, list_freqs):
    if eeg is None or list_freqs is None:
        raise ValueError('Not enough input arguments.')

    # Filter bank coefficients
    fb_coefs = np.array([i for i in range(1, fbcca_config['subBands'] + 1)])**(-1.25) + 0.25

    # Compute all sub-bands once per invocation to avoid redundant filtering
    filtered_subbands = [filterbank(eeg, fb_idx + 1) for fb_idx in range(fbcca_config['subBands'])]
    if not filtered_subbands:
        raise ValueError('No sub-bands computed; check fbcca_config.subBands.')

    num_smpls_resampled = filtered_subbands[0].shape[1]

    # Generate reference signals to match downsampled EEG length
    y_ref = cca_reference(list_freqs, num_smpls_resampled, fs=256)  # <-- updated

    r = np.zeros((fbcca_config['subBands'], len(list_freqs)))

    for fb_i, testdata in enumerate(filtered_subbands):
        for class_i in range(len(list_freqs)):
            refdata = y_ref[class_i, :, :]

            n_components = min(testdata.shape[0], refdata.shape[0])
            cca = CCA(n_components=n_components)
            cca.fit(testdata.T, refdata.T)

            testdata_c, refdata_c = cca.transform(testdata.T, refdata.T)
            r_tmp = [np.corrcoef(testdata_c[:, i], refdata_c[:, i])[0, 1] for i in range(n_components)]
            r[fb_i, class_i] = r_tmp[0]

    # Weighted sum of correlations
    rho = np.dot(fb_coefs, r)
    correlation = np.max(rho)
    tau = np.argmax(rho)

    if correlation < fbcca_config['correlationThreshold']:
        estimated_label = fbcca_config['idleStateLabel']
    else:
        estimated_label = tau

    return estimated_label


def cca_reference(list_freqs, num_smpls, fs=256):  # fs parameter added
    if list_freqs is None or num_smpls is None:
        raise ValueError('Not enough input arguments.')

    num_freqs = len(list_freqs)
    tidx = np.arange(1, num_smpls + 1) / fs  # <-- use fs, not fbcca_config['samplingRate']

    y_ref = np.zeros((num_freqs, 2 * fbcca_config['harmonics'], num_smpls))

    for freq_i in range(num_freqs):
        tmp = []
        for harm_i in range(1, fbcca_config['harmonics'] + 1):
            stim_freq = list_freqs[freq_i]
            tmp.append(np.sin(2 * np.pi * tidx * harm_i * stim_freq))
            tmp.append(np.cos(2 * np.pi * tidx * harm_i * stim_freq))
        y_ref[freq_i, :, :] = np.array(tmp)

    return y_ref